File: ios/Classes/TrustPinSDKPlugin.swift
import Flutter
import TrustPinKit
import UIKit

public class TrustPinSDKPlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "trustpin_sdk", binaryMessenger: registrar.messenger())
    let instance = TrustPinSDKPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "setup":
      handleSetup(call: call, result: result)
    case "verify":
      handleVerify(call: call, result: result)
    case "setLogLevel":
      handleSetLogLevel(call: call, result: result)
    default:
      Task { @MainActor in
        result(FlutterMethodNotImplemented)
      }
    }
  }

  private func handleSetup(call: FlutterMethodCall, result: @escaping FlutterResult) {
    Task {
      do {
        guard let args = call.arguments as? [String: Any],
              let organizationId = args["organizationId"] as? String,
              let projectId = args["projectId"] as? String,
              let publicKey = args["publicKey"] as? String
        else {
          await MainActor.run {
            result(
                FlutterError(
                    code: "INVALID_ARGUMENTS",
                    message: "Missing required arguments",
                    details: nil))
          }
          return
        }

        let configurationURL = args["configurationURL"] as? String
        let modeString = args["mode"] as? String ?? "strict"
        let mode: TrustPinMode = modeString == "permissive" ? .permissive : .strict

        if let configURLString = configurationURL, !configURLString.isEmpty,
           let url = URL(string: configURLString) {
          try await TrustPin.setup(
              organizationId: organizationId,
              projectId: projectId,
              publicKey: publicKey,
              configurationURL: url,
              mode: mode)
        } else {
          try await TrustPin.setup(
              organizationId: organizationId,
              projectId: projectId,
              publicKey: publicKey,
              mode: mode)
        }

        await MainActor.run {
          result(nil)
        }
      } catch let error as TrustPinErrors {
        await MainActor.run {
          result(
              FlutterError(
                  code: mapTrustPinError(error),
                  message: error.localizedDescription,
                  details: nil))
        }
      } catch {
        await MainActor.run {
          result(
              FlutterError(
                  code: "SETUP_ERROR",
                  message: error.localizedDescription,
                  details: nil))
        }
      }
    }
  }

  private func handleVerify(call: FlutterMethodCall, result: @escaping FlutterResult) {
    Task {
      do {
        guard let args = call.arguments as? [String: Any],
              let domain = args["domain"] as? String,
              let certificate = args["certificate"] as? String
        else {
          await MainActor.run {
            result(
                FlutterError(
                    code: "INVALID_ARGUMENTS",
                    message: "Missing required arguments",
                    details: nil))
          }
          return
        }

        try await TrustPin.verify(domain: domain, certificate: certificate)

        await MainActor.run {
          result(nil)
        }
      } catch let error as TrustPinErrors {
        await MainActor.run {
          result(
              FlutterError(
                  code: mapTrustPinError(error),
                  message: error.localizedDescription,
                  details: nil))
        }
      } catch {
        await MainActor.run {
          result(
              FlutterError(
                  code: "VERIFY_ERROR",
                  message: error.localizedDescription,
                  details: nil))
        }
      }
    }
  }

  private func handleSetLogLevel(call: FlutterMethodCall, result: @escaping FlutterResult) {
    Task {
      guard let args = call.arguments as? [String: Any],
            let logLevelString = args["logLevel"] as? String
      else {
        await MainActor.run {
          result(
              FlutterError(
                  code: "INVALID_ARGUMENTS",
                  message: "Missing logLevel argument",
                  details: nil))
        }
        return
      }

      let logLevel: TrustPinLogLevel
      switch logLevelString.lowercased() {
      case "none":
        logLevel = .none
      case "error":
        logLevel = .error
      case "info":
        logLevel = .info
      case "debug":
        logLevel = .debug
      default:
        logLevel = .error
      }

      await TrustPin.set(logLevel: logLevel)

      await MainActor.run {
        result(nil)
      }
    }
  }

  private func mapTrustPinError(_ error: TrustPinErrors) -> String {
    switch error {
    case .invalidProjectConfig:
      return "INVALID_PROJECT_CONFIG"
    case .errorFetchingPinningInfo:
      return "ERROR_FETCHING_PINNING_INFO"
    case .invalidServerCert:
      return "INVALID_SERVER_CERT"
    case .pinsMismatch:
      return "PINS_MISMATCH"
    case .allPinsExpired:
      return "ALL_PINS_EXPIRED"
    case .configurationValidationFailed:
      return "CONFIGURATION_VALIDATION_FAILED"
    case .domainNotRegistered:
      return "DOMAIN_NOT_REGISTERED"
    @unknown default:
      return "INVALID_PROJECT_CONFIG"
    }
  }
}

File: android/src/main/kotlin/cloud/trustpin/flutter/sdk/TrustPinSDKPlugin.kt
package cloud.trustpin.flutter.sdk

import cloud.trustpin.kotlin.sdk.TrustPin
import cloud.trustpin.kotlin.sdk.TrustPinError
import cloud.trustpin.kotlin.sdk.TrustPinLogLevel
import cloud.trustpin.kotlin.sdk.TrustPinMode

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

import java.security.cert.CertificateFactory
import java.security.cert.X509Certificate
import java.io.ByteArrayInputStream
import java.util.Base64
import java.net.URI
import java.net.URL

/** TrustPinSDKPlugin */
class TrustPinSDKPlugin: FlutterPlugin, MethodCallHandler {
  private lateinit var channel : MethodChannel
  private val coroutineScope = CoroutineScope(Dispatchers.Main)

  override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "trustpin_sdk")
    channel.setMethodCallHandler(this)
  }

  override fun onMethodCall(call: MethodCall, result: Result) {
    when (call.method) {
      "setup" -> handleSetup(call, result)
      "verify" -> handleVerify(call, result)
      "setLogLevel" -> handleSetLogLevel(call, result)
      else -> result.notImplemented()
    }
  }

  private fun handleSetup(call: MethodCall, result: Result) {
    coroutineScope.launch {
      try {
        val organizationId = call.argument<String>("organizationId")
        val projectId = call.argument<String>("projectId")
        val publicKey = call.argument<String>("publicKey")
        val configurationURL = call.argument<String>("configurationURL")
        val modeString = call.argument<String>("mode") ?: "strict"

        if (organizationId == null || projectId == null || publicKey == null) {
          result.error("INVALID_ARGUMENTS", "Missing required arguments", null)
          return@launch
        }

        val mode = when (modeString.lowercase()) {
          "permissive" -> TrustPinMode.PERMISSIVE
          "strict" -> TrustPinMode.STRICT
          else -> TrustPinMode.STRICT
        }

        if (configurationURL != null && configurationURL.isNotEmpty()) {
          val url = try {
            URI.create(configurationURL).toURL()
          } catch (e: Exception) {
            throw TrustPinError.InvalidProjectConfig
          }

          TrustPin.setup(organizationId, projectId, publicKey, url, mode)
        } else {
          TrustPin.setup(organizationId, projectId, publicKey, mode)
        }

        result.success(null)
      } catch (e: TrustPinError) {
        result.error(mapTrustPinError(e), e.message, null)
      } catch (e: Exception) {
        result.error("SETUP_ERROR", e.message, null)
      }
    }
  }

  private fun handleVerify(call: MethodCall, result: Result) {
    coroutineScope.launch {
      try {
        val domain = call.argument<String>("domain")
        val certificatePem = call.argument<String>("certificate")

        if (domain == null || certificatePem == null) {
          result.error("INVALID_ARGUMENTS", "Missing required arguments", null)
          return@launch
        }

        val certificate = parsePemCertificate(certificatePem)
        TrustPin.verify(domain, certificate)
        result.success(null)
      } catch (e: TrustPinError) {
        result.error(mapTrustPinError(e), e.message, null)
      } catch (e: Exception) {
        result.error("VERIFY_ERROR", e.message, null)
      }
    }
  }

  private fun handleSetLogLevel(call: MethodCall, result: Result) {
    try {
      val logLevelString = call.argument<String>("logLevel")
      if (logLevelString == null) {
        result.error("INVALID_ARGUMENTS", "Missing logLevel argument", null)
        return
      }

      val logLevel = when (logLevelString.lowercase()) {
        "none" -> TrustPinLogLevel.NONE
        "error" -> TrustPinLogLevel.ERROR
        "info" -> TrustPinLogLevel.INFO
        "debug" -> TrustPinLogLevel.DEBUG
        else -> TrustPinLogLevel.ERROR
      }

      TrustPin.setLogLevel(logLevel)
      result.success(null)
    } catch (e: Exception) {
      result.error("SET_LOG_LEVEL_ERROR", e.message, null)
    }
  }

  private fun parsePemCertificate(pemCertificate: String): X509Certificate {
    val certificateFactory = CertificateFactory.getInstance("X.509")

    // Extract the certificate content between BEGIN and END markers
    val cleanPem = pemCertificate
      .replace("-----BEGIN CERTIFICATE-----", "")
      .replace("-----END CERTIFICATE-----", "")
      .replace("\\s".toRegex(), "")

    val decodedBytes = Base64.getDecoder().decode(cleanPem)
    val inputStream = ByteArrayInputStream(decodedBytes)

    return certificateFactory.generateCertificate(inputStream) as X509Certificate
  }

  private fun mapTrustPinError(error: TrustPinError): String {
    return when (error) {
      is TrustPinError.InvalidProjectConfig -> "INVALID_PROJECT_CONFIG"
      is TrustPinError.ErrorFetchingPinningInfo -> "ERROR_FETCHING_PINNING_INFO"
      is TrustPinError.InvalidServerCert -> "INVALID_SERVER_CERT"
      is TrustPinError.PinsMismatch -> "PINS_MISMATCH"
      is TrustPinError.AllPinsExpired -> "ALL_PINS_EXPIRED"
      is TrustPinError.ConfigurationValidationFailed -> "CONFIGURATION_VALIDATION_FAILED"
      is TrustPinError.DomainNotRegistered -> "DOMAIN_NOT_REGISTERED"
      else -> "UNKNOWN_ERROR"
    }
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }
}